test/at_coroutine_exit_test.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <unifex/coroutine.hpp>
+
+#if !UNIFEX_NO_COROUTINES
+
+#include <unifex/at_coroutine_exit.hpp>
+
+#include <unifex/task.hpp>
+#include <unifex/stop_if_requested.hpp>
+#include <unifex/sync_wait.hpp>
+
+#include <gtest/gtest.h>
+
+using namespace unifex;
+
+namespace {
+int global = 0;
+
+task<void> test_one_cleanup_action() {
+  ++global;
+  co_await at_coroutine_exit([]() -> task<void> { global *= 2; co_return; });
+  ++global;
+}
+
+task<void> test_two_cleanup_actions() {
+  ++global;
+  co_await at_coroutine_exit([]() -> task<void> { global *= 2; co_return; });
+  co_await at_coroutine_exit([]() -> task<void> { global *= global; co_return; });
+  ++global;
+}
+
+task<void> test_one_cleanup_action_with_stop() {
+  ++global;
+  co_await at_coroutine_exit([]() -> task<void> { global *= 2; co_return; });
+  co_await stop();
+  ++global;
+}
+
+task<void> test_two_cleanup_actions_with_stop() {
+  ++global;
+  co_await at_coroutine_exit([]() -> task<void> { global *= 2; co_return; });
+  co_await at_coroutine_exit([]() -> task<void> { global += global; co_return; });
+  co_await stop();
+  ++global;
+}
+
+task<void> with_continuation(unifex::task<void> next) {
+  co_await std::move(next);
+  global *= 3;
+}
+
+}
+
```

Comment:
Per [the docs on streams](https://github.com/facebookexperimental/libunifex/blob/master/doc/concepts.md#streams), a stream's clean-up action must complete with `set_done` (if clean-up succeeded) or `set_error` (if not).  We won't be able to handle streams in the way I expected to if cancelled clean-up actions lead to termination.  This will fail at runtime:
```cpp
task<void> for_each(Stream auto stream, Callable auto func) {
  auto& [stream] = at_coroutine_exit([](auto& stream) -> task<void> {
    // this will terminate if it behaves as it's spec'd to
    co_await cleanup(stream);
  }, std::move(stream));

  for (;;) {
    func(co_await next(stream));
  }
}
```